vtex-apps/admin-catalog-translations
# Increase download limit

**What problem is this solving?**


This PR removes the limit of 1600 when creating a file to download product translations. It adds a tab that renders all translation requests, and shows the current status of a file: `processing`, `error` or ready to `downalod`. 

When requesting a new translation, a new line will be added to the panel. If the processing time is too long, user can check the translation out in a latter time. When download is ready, user can clicks the button that will trigger the download.


#### Back end
For the back end, it adds 4 new resolvers:

__productTranslations__:  Used to trigger a translation process. It will calls a function that resolves after the server respond to the client. It updates an object on VBASE that holds all `requestId` in an array, creates an object associated - below - to this `requestId` and sends it to the client as a response.

```ts
interface ProductTranslationRequest {
  requestId: string
  estimatedTime: number // Calculate by multiplying the number of products per the `sleep time` for every call to translate products
  requestedBy: string
  categoryId: string
  error?: boolean
  createdAt: Date
  locale: string
  completedAt?: Date // Update only when the async function translating the products finishes
  translations?: Array<GraphQLResponse<ProductTranslationResponse>> | null // Update only when the async function translating the products finishes
}
```

__productTranslationRequests__: Used to access all requests made to translate products. It holds only the `requestId` in an array.

__productTranslationRequestInfo__: Used to consult the object with the summary of the translation process (above).

__downloadProductTranslation__: Used to resolve the product translation info. Client calls it when the the field `completedAt` has some value.


#### Front end

For the front end, it removes the modal for export from the main component `react/components/ProductTranslation/ProductTranslation.tsx` and brings it to `ProductExportModal.tsx`. 

The best order to review the front end files is: `ProductExportModal.tsx` then `ExportListItem.tsx`

__ProductExportModal.tsx__

In this component is where the trigger to start the process to translate lives. 

The [lazyQuery](https://github.com/vtex-apps/admin-catalog-translations/compare/increase-download-limit?expand=1#diff-8e0898cdd28bea4909f29e0088e1c07c0711f40b3f9e4299e78de57671e8a0d1R55-R67) triggers the translation process, given a locale and category id. 

And this [query](https://github.com/vtex-apps/admin-catalog-translations/compare/increase-download-limit?expand=1#diff-8e0898cdd28bea4909f29e0088e1c07c0711f40b3f9e4299e78de57671e8a0d1R69-R71) fetches all translation requests, used to create the list of files rendered by `ExportListItem.tsx`.

The main [useEffect](https://github.com/vtex-apps/admin-catalog-translations/compare/increase-download-limit?expand=1#diff-8e0898cdd28bea4909f29e0088e1c07c0711f40b3f9e4299e78de57671e8a0d1R73-R89) runs after a request is made and updates the query result that contains all requests id with the new one created. 

__ExportListItem.tsx__

Every request id renders this component. In the first render, it fetches the details for the request [here](https://github.com/vtex-apps/admin-catalog-translations/compare/increase-download-limit?expand=1#diff-9c0312ec6eeb61f6f02602aec03d7c8734c4ae3dc604cae5f59d1b7877d36721R18-R28)

The main [useEffect](https://github.com/vtex-apps/admin-catalog-translations/compare/increase-download-limit?expand=1#diff-9c0312ec6eeb61f6f02602aec03d7c8734c4ae3dc604cae5f59d1b7877d36721R55-R75) is responsible for the logic the start and stop polling for new info when the request detail doesn't have either the `completedAt` or `error` flags. It relays in the `estimation` time to set the polling interval, to calculate if the translations should already have finished and the remaining time until finishing the process. For the latter two, it adds a margin to avoid false negatives.

The small [useEffect](https://github.com/vtex-apps/admin-catalog-translations/compare/increase-download-limit?expand=1#diff-9c0312ec6eeb61f6f02602aec03d7c8734c4ae3dc604cae5f59d1b7877d36721R55-R75) refetches the query if use closes and opens again the modal and the [last one](https://github.com/vtex-apps/admin-catalog-translations/compare/increase-download-limit?expand=1#diff-9c0312ec6eeb61f6f02602aec03d7c8734c4ae3dc604cae5f59d1b7877d36721R77-R86) handles the parse process from json to xlsx after it is downloaded with this [lazyQuery](https://github.com/vtex-apps/admin-catalog-translations/compare/increase-download-limit?expand=1#diff-9c0312ec6eeb61f6f02602aec03d7c8734c4ae3dc604cae5f59d1b7877d36721R30-R35).

**How should this be manually tested?**

1. Visit [this workspace](https://exportproducts--powerplanet.myvtex.com/admin/catalog-translation/product);
2. Select a language different then the main one;
3. Open the `export` modal;
4. Select a category;
5. Click export products;

There should be a new line in the tab `See Files`. If it's a small category (category 222), the file should be ready to be downloaded. If it's a big category (334), the process should take ~3 minutes. One can either wait in the page or close it and check it later.

**Screenshots or example usage:**

<img width="838" alt="Screen Shot 2021-05-06 at 5 26 27 PM" src="https://user-images.githubusercontent.com/38737958/117324609-3c973c00-ae90-11eb-9c96-4f79b55679d0.png">
